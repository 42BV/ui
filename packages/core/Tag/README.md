# @42.nl/ui-core-tag

## Installation

Install the package in your project directory with:

```sh
// with npm
npm install @42.nl/ui-core-tag

// with yarn
yarn add @42.nl/ui-core-tag
```
