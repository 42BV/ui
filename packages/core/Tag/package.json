{
  "name": "@42.nl/ui-core-tag",
  "version": "0.0.1",
  "description": "",
  "main": "lib/Tag.js",
  "private": false,
  "scripts": {
    "test": "jest --config=jest.config.js",
    "build": "tsc",
    "lint": "eslint src --ext js,jsx,ts,tsx --max-warnings=0",
    "clean": "rm -rf lib"
  },
  "files": [
    "lib/*"
  ],
  "dependencies": {
    "classnames": "2.2.6"
  },
  "peerDependencies": {
    "react": "^16.8.0",
    "react-dom": "^16.8.0",
    "reactstrap": "^8.0.1"
  },
  "devDependencies": {
    "reactstrap": "8.0.1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "publishConfig": {
    "access": "public"
  }
}
