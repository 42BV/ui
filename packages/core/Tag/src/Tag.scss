@import '~@42.nl/ui-core-styling/dist/scss/abstracts/variables';

.token {
  &.token-info {
    background-color: $teal !important;
  }

  &.token-success {
    background-color: $green !important;
  }

  &.token-warning {
    background-color: $orange !important;
  }

  &.token-danger {
    background-color: $red !important;
  }

  .close-button {
    cursor: pointer;
  }
}
