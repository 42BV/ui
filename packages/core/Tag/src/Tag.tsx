import React from 'react';
import classNames from 'classnames';
import { Color } from '@42.nl/ui-types';

type Type = 'info' | 'success' | 'warning' | 'danger';

interface Props {
  /**
   * Text to display within the tag.
   */
  text: string;

  /**
   * The color of the badge. Supports Bootstrap colors (e.g. primary, danger).
   */
  color?: Color;

  /**
   * Determines if a tag should be removable.
   */
  onRemove?: () => void;

  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   */
  className?: string;
}

export default function Tag({ text, color, onRemove, className }: Props) {
  const canClose = onRemove !== undefined;
  const classes = classNames('token', className, {
    'token-removeable': canClose,
    [`token-${color}`]: color
  });

  return (
    <span className="tag">
      <span className={classes}>
        {text}

        {canClose ? (
          <span onClick={onRemove} className="close-button" role="button">
            Ã—
          </span>
        ) : null}
      </span>
    </span>
  );
}
