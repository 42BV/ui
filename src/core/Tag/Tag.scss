.badge {
  display: inline-block;
  margin-right: 0.1875rem;
  cursor: default;

  .close-button {
    cursor: pointer;
    margin-left: 0.1875rem;
  }
}
