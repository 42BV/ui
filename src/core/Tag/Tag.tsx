import React from 'react';
import classNames from 'classnames';
import { Color } from '../types';

type Props = {
  /**
   * Text to display within the tag.
   */
  text: string;

  /**
   * The color of the badge. Supports Bootstrap colors (e.g. primary, danger).
   */
  color?: Color;

  /**
   * Determines if a tag should be removable.
   */
  onRemove?: () => void;

  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   */
  className?: string;
};

/**
 * Tag is a component which shows a piece of text inside of a colored
 * background. Under the hood the Tag component is a bootstrap badge
 * with the option to show a close button.
 *
 * An example of a good use case for using a `Tag` is for blog post
 * which belong in a category, each category could be a `Tag`. Another
 * example is someone' hobby's on a profile page, each hobby could
 * be a `Tag`.
 */
export default function Tag({ text, color, onRemove, className }: Props) {
  const canClose = onRemove !== undefined;
  const classes = classNames('tag', className, {
    [`tag-${color}`]: color
  });

  return (
    <span className={classes}>
      {text}

      {canClose ? (
        <span onClick={onRemove} className="close-button" role="button">
          Ã—
        </span>
      ) : null}
    </span>
  );
}
