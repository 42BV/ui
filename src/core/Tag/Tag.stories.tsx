import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import { Tag } from './Tag';

storiesOf('core/Tag', module)
  .addParameters({ component: Tag })
  .add('default', () => (
    <div className="text-center">
      <Tag text="Maarten" />
      <Tag text="Maarten" color="success" />
      <Tag text="Maarten" color="info" />
      <Tag text="Maarten" color="warning" />
      <Tag text="Maarten" color="danger" />
      <Tag text="Maarten" onRemove={action('on close')} />
    </div>
  ));
