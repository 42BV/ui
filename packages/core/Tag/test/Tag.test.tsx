import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Tag from '../src/Tag';

describe('Component: Tag', () => {
  describe('ui', () => {
    test('default', () => {
      const tag = shallow(<Tag text="Maarten" />);
      expect(toJson(tag)).toMatchSnapshot('Component: Tag => ui => default');
    });

    test('with type', () => {
      const tag = shallow(<Tag text="Maarten" color="success" />);
      expect(toJson(tag)).toMatchSnapshot('Component: Tag => ui => with type');
    });

    test('with close', () => {
      const tag = shallow(<Tag text="Maarten" onRemove={() => undefined} />);
      expect(toJson(tag)).toMatchSnapshot('Component: Tag => ui => with close');
    });
  });

  test('onClose', () => {
    const onCloseSpy = jest.fn();
    const tag = shallow(<Tag text="Maarten" onRemove={onCloseSpy} />);

    tag.find('span.close-button').simulate('click');
    expect(onCloseSpy).toHaveBeenCalledTimes(1);
  });
});
