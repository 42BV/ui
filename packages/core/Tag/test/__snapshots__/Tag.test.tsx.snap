// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Tag ui default: Component: Tag => ui => default 1`] = `
<span
  className="tag"
>
  <span
    className="token"
  >
    Maarten
  </span>
</span>
`;

exports[`Component: Tag ui with close: Component: Tag => ui => with close 1`] = `
<span
  className="tag"
>
  <span
    className="token token-removeable"
  >
    Maarten
    <span
      className="close-button"
      onClick={[Function]}
      role="button"
    >
      Ã—
    </span>
  </span>
</span>
`;

exports[`Component: Tag ui with type: Component: Tag => ui => with type 1`] = `
<span
  className="tag"
>
  <span
    className="token token-success"
  >
    Maarten
  </span>
</span>
`;
